packages/opencensus-instrumentation-http2/src/http2.ts

```diff
@@ -0,0 +1,283 @@
+    }
+  }
+
+  patchConnect() {
+    const plugin = this;
+    return (original: ConnectFunction):
+               types.Func<http2.ClientHttp2Session> => {
+      return function patchedConnect(authority: string):
+          http2.ClientHttp2Session {
+            plugin.client = original.apply(this, arguments);
+            shimmer.wrap(plugin.client, 'request', plugin.patchRequest());
+
+            shimmer.unwrap(plugin.moduleExports, 'connect');
+            plugin.patchedConnect = true;
+            plugin.clientUrl = url.parse(authority);
+
+            return plugin.client;
+          };
+    };
+  }
+
+  patchRequest() {
+    const plugin = this;
+    return (original: RequestFunction): types.Func<http2.ClientHttp2Stream> => {
+      return function patchedRequest(
+                 headers: http2.IncomingHttpHeaders): http2.ClientHttp2Stream {
+        // Do not trace ourselves
+        if (headers['x-opencensus-outgoing-request']) {
+          return original.apply(this, arguments);
+        }
+
+        const request = original.apply(this, arguments);
+        plugin.tracer.wrapEmitter(request);
+
+        const traceOptions = {
+          name: `${headers[':method'] ? headers[':method'] : 'GET'} ${
+              headers[':path']}`,
+          type: 'CLIENT',
+        };
+
+        // Checks if this outgoing request is part of an operation by checking
+        // if there is a current root span, if so, we create a child span. In
+        // case there is no root span, this means that the outgoing request is
+        // the first operation, therefore we create a root span.
+        if (!plugin.tracer.currentRootSpan) {
+          return plugin.tracer.startRootSpan(
+              traceOptions,
+              plugin.makeHttp2RequestTrace(request, headers, plugin));
+        } else {
+          const span = plugin.tracer.startChildSpan(
+              traceOptions.name, traceOptions.type);
+          return (plugin.makeHttp2RequestTrace(request, headers, plugin))(span);
+        }
+      };
+    };
+  }
+
+  makeHttp2RequestTrace(
+      request: http2.ClientHttp2Stream, headers: http2.IncomingHttpHeaders,
+      plugin: Http2Plugin): types.Func<http2.ClientHttp2Stream> {
+    return (span: types.Span): http2.ClientHttp2Stream => {
+      if (!span) return request;
+
+      const setter = {
+        setHeader(name: string, value: string) {
+          headers[name] = value;
+        }
+      } as types.HeaderSetter;
+
+      const propagation = plugin.tracer.propagation;
+      if (propagation) {
+        propagation.inject(setter, span.spanContext);
+      }
+
+      request.on('response', (responseHeaders: http2.IncomingHttpHeaders) => {
+        span.addAttribute(
+            Http2Plugin.ATTRIBUTE_HTTP_STATUS_CODE,
+            `${responseHeaders[':status']}`);
+        span.status = plugin.traceStatus(+responseHeaders[':status']);
+      });
+
+      request.on('end', () => {
+        const userAgent =
```

Comment:
Should this be `string|null`? Then wrap the `addAttribute` for user agent in a conditional.

Reply:
It works, just changed, but `userAgent` comes as `string | string[]`, so I had to cast to `string` before passing as attribute.

Reply:
Ok, makes sense. I forgot that outgoing headers can be `string[]` as well.